### PR TITLE
decouple gc worker from Storage

### LINKED ISSUE #5398: decouple gc worker from Storage
## Question

I find that the Storage only bypass `async_gc` function to GC worker for outside GC, so I think it is better to decouple GC worker from Storage and let outside call GC worker directly. 

This can simplify Storage dependence because GC worker depends on raftstore. Maybe this can decrease compiling time too.

### PR DESCRIPTION
**Pull request by FateTHarlaown**
Signed-off-by: FateTHarlaown <623239185@qq.com>

###  What have you changed?

Please explain in detail what the changes are in this PR and why they are needed:

- Summarize your change
decouple gc worker from Storag ,  for #5398
- How does this PR work? Include a brief introduction for any changed logic (optional)
moved the `src/storage/gc_worker.rs` to `src/server/gc_worker.rs`,  passed a `GCWorker` to `Service` to call and deleted the gc interface of `Storage`.

###  What is the type of the changes?
- Improvement (a change which is an improvement to an existing feature)

###  How is the PR tested?
Please select the tests that you ran to verify your changes:
- Unit test
- Integration test

###  Refer to a related PR or issue link (optional)
Fixes #5398

### CODE REVIEW

**Comment by siddontang** [comment]
Cool @FateTHarlaown 

PTAL @MyonKeminta

File: src/server/service/kv.rs [hunk 1]
```diff
@@ -611,12 +617,13 @@ impl<T: RaftStoreRouter + 'static, E: Engine, L: LockMgr> Tikv for Service<T, E,
...
 
         let (cb, f) = paired_future_callback();
-        let res = self.storage.async_unsafe_destroy_range(
+        let res = self.gc_worker.async_unsafe_destroy_range(
             req.take_context(),
             Key::from_raw(&req.take_start_key()),
             Key::from_raw(&req.take_end_key()),
             cb,
         );
+        KV_COMMAND_COUNTER_VEC_STATIC.gc.inc();
```
Comment (MyonKeminta) [inline src/server/service/kv.rs]: I think you'd better put this line (and also that in `future_gc`) in `GCWorker`'s methods, which is just like those storage commands updates this metrics in `Storage`'s methods. By the way, it seems you updated wrong metric here.

File: cmd/src/server.rs [hunk 2]
```diff
@@ -186,12 +187,20 @@ fn run_raft_server(pd_client: RpcClient, cfg: &TiKvConfig, security_mgr: Arc<Sec
         None
     };
 
+    let mut gc_worker = GCWorker::new(
```
Comment (youjiali1995) [inline cmd/src/server.rs]: `GCWorker` and `GCManager` can start together. It's not related to this PR. Maybe we can extract a single `start` method later.

File: components/test_raftstore/src/server.rs [hunk 3]
```diff
@@ -138,14 +139,14 @@ impl Simulator for ServerCluster {
...
-            None,
-            None,
-            None,
-        )?;
+
+        let engine = RaftKv::new(sim_router.clone());
+
+        let mut gc_worker =
+            GCWorker::new(engine.clone(), None, None, cfg.storage.gc_ratio_threshold);
+        gc_worker.start().unwrap();
```
Comment (youjiali1995) [inline components/test_raftstore/src/server.rs]: Is it unnecessary? @MyonKeminta

Comment (MyonKeminta) [reply components/test_raftstore/src/server.rs]: Of cource we can make it RAII, just like how `storage` does. I think it's a good idea.

File: src/server/gc_worker.rs [hunk 4]
```diff
@@ -1059,12 +1058,49 @@ pub struct GCWorker<E: Engine> {
...
+    /// How many strong references. The worker will be stopped
+    /// once there are no more references.
+    refs: Arc<atomic::AtomicUsize>,
     worker: Arc<Mutex<Worker<GCTask>>>,
     worker_scheduler: worker::Scheduler<GCTask>,
 
     gc_manager_handle: Arc<Mutex<Option<GCManagerHandle>>>,
 }
 
+impl<E: Engine> Clone for GCWorker<E> {
```
Comment (youjiali1995) [inline src/server/gc_worker.rs]: Now the `GCWoker` starts explicitly but stops implicitly. It's weird to me. Maybe we can call `GCWorker.stop()` in `run_raft_server()` like other workers.

Comment (FateTHarlaown) [reply src/server/gc_worker.rs]: how about creating a `GCWorkerBuilder` and call `start` in its `build` function ? So, `GCWoker` can be started and stoped implicitly .

**Pull request merged**